File: src/EG016SetTabValues.php
<?php
/**
 * Example 016: Set optional and locked field values and an envelope custom field value
 */

namespace Example;
class EG016SetTabValues
{

    private $eg = "eg016";  # reference (and url) for this example
    private $signer_client_id = 1000; # Used to indicate that the signer will use an embedded
                # Signing Ceremony. Represents the signer's userId within your application.

    public function controller()
    {
        $method = $_SERVER['REQUEST_METHOD'];
        if ($method == 'GET') {
            $this->getController();
        };
        if ($method == 'POST') {
            check_csrf();
            $this->createController();
        };
    }

    /**
     * 1. Check the token
     * 2. Call the worker method
     * 3. Redirect the user to the signing ceremony
     */
    private function createController()
    {
        $minimum_buffer_min = 3;
        if (ds_token_ok($minimum_buffer_min)) {
            # 2. Call the worker method
            # More data validation would be a good idea here
            # Strip anything other than characters listed

            $signer_name = preg_replace('/([^\w \-\@\.\,])+/', '', $_POST['signer_name']);
            $signer_email = preg_replace('/([^\w \-\@\.\,])+/', '', $_POST['signer_email']);
            $envelope_args = [
                'signer_email' => $signer_email,
                'signer_name' => $signer_name,
                'signer_client_id' => $this->signer_client_id,
                'ds_return_url' => $GLOBALS['app_url'] . 'index.php?page=ds_return'
            ];
            $args = [
                'account_id' => $_SESSION['ds_account_id'],
                'base_path' => $_SESSION['ds_base_path'],
                'ds_access_token' => $_SESSION['ds_access_token'],
                'envelope_args' => $envelope_args
            ];

            try {
                $results = $this->worker($args);
            } catch (\DocuSign\eSign\ApiException $e) {
                $error_code = $e->getResponseBody()->errorCode;
                $error_message = $e->getResponseBody()->message;
                $GLOBALS['twig']->display('error.html', [
                        'error_code' => $error_code,
                        'error_message' => $error_message]
                );
                exit();
            }
            if ($results) {
                $_SESSION["envelope_id"] = $results["envelope_id"]; # Save for use by other examples
                                                                    # which need an envelopeId

                # Redirect the user to the Signing Ceremony
                # Don't use an iFrame!
                # State can be stored/recovered using the framework's session or a
                # query parameter on the returnUrl (see the makeRecipientViewRequest method)
                header('Location: ' . $results["redirect_url"]);
                exit;
            }
        } else {
            flash('Sorry, you need to re-authenticate.');
            # We could store the parameters of the requested operation
            # so it could be restarted automatically.
            # But since it should be rare to have a token issue here,
            # we'll make the user re-enter the form data after
            # authentication.
            $_SESSION['eg'] = $GLOBALS['app_url'] . 'index.php?page=' . $this->eg;
            header('Location: ' . $GLOBALS['app_url'] . 'index.php?page=must_authenticate');
            exit;
        }
    }


    /**
     * Do the work of the example
     * 1. Create the envelope request object
     * 2. Send the envelope
     * 3. Create the Recipient View request object
     * 4. Obtain the recipient_view_url for the signing ceremony
     * @param $args
     * @return array ['redirect_url']
     * @throws \DocuSign\eSign\ApiException for API problems and perhaps file access \Exception too.
     */
    # ***DS.snippet.0.start
    private function worker($args)
    {
        $envelope_args = $args["envelope_args"];
        # 1. Create the envelope request object
        $envelope_definition = $this->make_envelope($envelope_args);

        # 2. call Envelopes::create API method
        # Exceptions will be caught by the calling function
        $config = new \DocuSign\eSign\Configuration();
        $config->setHost($args['base_path']);
        $config->addDefaultHeader('Authorization', 'Bearer ' . $args['ds_access_token']);
        $api_client = new \DocuSign\eSign\ApiClient($config);
        $envelope_api = new \DocuSign\eSign\Api\EnvelopesApi($api_client);
        $results = $envelope_api->createEnvelope($args['account_id'], $envelope_definition);
        $envelope_id = $results->getEnvelopeId();

        # 3. Create the Recipient View request object
        $authentication_method = 'None'; # How is this application authenticating
        # the signer? See the `authenticationMethod' definition
        # https://developers.docusign.com/esign-rest-api/reference/Envelopes/EnvelopeViews/createRecipient

        $recipient_view_request = new \DocuSign\eSign\Model\RecipientViewRequest([
            'authentication_method' => $authentication_method,
            'client_user_id' => $envelope_args['signer_client_id'],
            'recipient_id' => '1',
            'return_url' => $envelope_args['ds_return_url'],
            'user_name' => $envelope_args['signer_name'], 'email' => $envelope_args['signer_email']
        ]);
        # 4. Obtain the recipient_view_url for the signing ceremony
        # Exceptions will be caught by the calling function
        $results = $envelope_api->createRecipientView($args['account_id'], $envelope_id,
            $recipient_view_request);

        return ['envelope_id' => $envelope_id, 'redirect_url' => $results['url']];
    }

    /**
     *  Creates envelope definition
     * @param $args parameters for the envelope:
     *              signer_email, signer_name, signer_client_id
     * @return mixed -- returns an envelope definition
     */
    private function make_envelope($args)
    {
        # document 1 (pdf) has tags
        # /sn1/ - signature field
        # /salary/ - yearly salary
        # /legal/ - legal name
        # /familiar/ - person's familiar name
        #
        # The envelope has one recipient.
        # recipient 1 - signer
        #
        # The salary is set both as a readable number in the
        # /salary/ text field, and as a pure number in a
        # custom field ('salary') in the envelope.

        # Salary that will be used.
        $salary = 123000;

        # Read the file
        $demo_docs_path = __DIR__ . '/../public/demo_documents/';
        $doc_name = 'World_Wide_Corp_salary.docx';
        $content_bytes = file_get_contents($demo_docs_path . $doc_name);
        $base64_file_content = base64_encode($content_bytes);

        # Create the document model
        $document = new \DocuSign\eSign\Model\Document([ # create the DocuSign document object
            'document_base64' => $base64_file_content,
            'name' => 'Salary action', # can be different from actual file name
            'file_extension' => 'docx', # many different document types are accepted
            'document_id' => 1 # a label used to reference the doc
        ]);

        # Create the signer recipient model
        $signer = new \DocuSign\eSign\Model\Signer([ # The signer
            'email' => $args['signer_email'], 'name' => $args['signer_name'],
            'recipient_id' => "1", 'routing_order' => "1",
            # Setting the client_user_id marks the signer as embedded
            'client_user_id' => $args['signer_client_id']
        ]);

        # Create a sign_here tab (field on the document)
        $sign_here = new \DocuSign\eSign\Model\SignHere([ # DocuSign SignHere field/tab
            'anchor_string' => '/sn1/', 'anchor_units' => 'pixels',
            'anchor_y_offset' => '10', 'anchor_x_offset' => '20'
        ]);

        # Create the legal and familiar text fields.
        # Recipients can update these values if they wish to.
        $text_legal = new \DocuSign\eSign\Model\Text([
            'anchor_string' => '/legal/', 'anchor_units' => 'pixels',
            'anchor_y_offset' => '-9', 'anchor_x_offset' => '5',
            'font' => "helvetica", 'font_size' => "size11",
            'bold' => 'true', 'value' => $args['signer_name'],
            'locked' => 'false', 'tab_id' => 'legal_name',
            'tab_label' => 'Legal name']);
        $text_familiar = new \DocuSign\eSign\Model\Text([
            'anchor_string' => '/familiar/', 'anchor_units' => 'pixels',
            'anchor_y_offset' => '-9', 'anchor_x_offset' => '5',
            'font' => "helvetica", 'font_size' => "size11",
            'bold' => 'true', 'value' => $args['signer_name'],
            'locked' => 'false', 'tab_id' => 'familiar_name',
            'tab_label' => 'Familiar name']);

        # Create the salary field. It should be human readable, so
        # add a comma before the thousands number, a currency indicator, etc.
        $salary_readable = '$' . number_format($salary);
        $text_salary = new \DocuSign\eSign\Model\Text([
            'anchor_string' => '/salary/', 'anchor_units' => 'pixels',
            'anchor_y_offset' => '-9', 'anchor_x_offset' => '5',
            'font' => "helvetica", 'font_size' => "size11",
            'bold' => 'true', 'value' => $salary_readable,
            'locked' => 'true', # mark the field as readonly
            'tab_id' => 'salary', 'tab_label' => 'Salary'
        ]);

        # Add the tabs model (including the sign_here tab) to the signer
        # The Tabs object wants arrays of the different field/tab types
        $signer->settabs(new \DocuSign\eSign\Model\Tabs(
            ['sign_here_tabs' => [$sign_here],
            'text_tabs' => [$text_legal, $text_familiar, $text_salary]]));

        # Create an envelope custom field to save the "real" (numeric)
        # version of the salary
        $salary_custom_field = new \DocuSign\eSign\Model\TextCustomField([
            'name' => 'salary',
            'required' => 'false',
            'show' => 'true', # Yes, include in the CoC
            'value' => $salary]);
        $custom_fields = new \DocuSign\eSign\Model\CustomFields([
            'text_custom_fields' => [$salary_custom_field]]);

        # Next, create the top level envelope definition and populate it.
        $envelope_definition = new \DocuSign\eSign\Model\EnvelopeDefinition([
            'email_subject' => "Please sign this document sent from the PHP SDK",
            'documents' => [$document],
            # The Recipients object wants arrays for each recipient type
            'recipients' => new \DocuSign\eSign\Model\Recipients(['signers' => [$signer]]),
            'status' => "sent", # requests that the envelope be created and sent.
            'custom_fields' => $custom_fields
        ]);

        return $envelope_definition;
    }
    # ***DS.snippet.0.end

    /**
     * Show the example's form page
     */
    private function getController()
    {
        if (ds_token_ok()) {
            $basename = basename(__FILE__);
            $GLOBALS['twig']->display('eg016_set_tab_values.html', [
                'title' => "Set field values",
                'source_file' => $basename,
                'source_url' => $GLOBALS['DS_CONFIG']['github_example_url'] . $basename,
                'documentation' => $GLOBALS['DS_CONFIG']['documentation'] . $this->eg,
                'show_doc' => $GLOBALS['DS_CONFIG']['documentation'],
                'signer_name' => $GLOBALS['DS_CONFIG']['signer_name'],
                'signer_email' => $GLOBALS['DS_CONFIG']['signer_email']
            ]);
        } else {
            # Save the current operation so it will be resumed after authentication
            $_SESSION['eg'] = $GLOBALS['app_url'] . 'index.php?page=' . $this->eg;
            header('Location: ' . $GLOBALS['app_url'] . 'index.php?page=must_authenticate');
            exit;
        }
    }
}

File: app/eg016_set_tab_values.py
"""Example 016: Set Tab Values"""

from flask import render_template, url_for, redirect, session, flash, request
from os import path
import json
from app import app, ds_config, views
import base64
import re
from docusign_esign import *
from docusign_esign.rest import ApiException

eg = "eg016"  # Reference (and URL) for this example
signer_client_id = 1000 # Used to indicate that the signer will use an embedded
                        # signing ceremony. Represents the signer's user ID within
                        # your application.
authentication_method = "None" # How is this application authenticating
                               # the signer? See the "authenticationMethod" definition
                               # https://developers.docusign.com/esign-rest-api/reference/Envelopes/EnvelopeViews/createRecipient

demo_docs_path = path.abspath(path.join(path.dirname(path.realpath(__file__)), "static/demo_documents"))

def controller():
    """Controller router using the HTTP method"""
    if request.method == "GET":
        return get_controller()
    elif request.method == "POST":
        return create_controller()
    else:
        return render_template("404.html"), 404

def create_controller():
    """
    1. Check the token
    2. Call the worker method
    3. Redirect the user to the signing ceremony
    """
    minimum_buffer_min = 3
    if views.ds_token_ok(minimum_buffer_min):
        # 2. Call the worker method
        # More data validation would be a good idea here
        # Strip anything other than characters listed
        pattern = re.compile("([^\w \-\@\.\,])+")
        signer_email = pattern.sub("", request.form.get("signer_email"))
        signer_name  = pattern.sub("", request.form.get("signer_name"))
        envelope_args = {
            "signer_email": signer_email,
            "signer_name": signer_name,
            "signer_client_id": signer_client_id,
            "ds_return_url": url_for("ds_return", _external=True),
        }
        args = {
            "account_id": session["ds_account_id"],
            "base_path": session["ds_base_path"],
            "ds_access_token": session["ds_access_token"],
            "envelope_args": envelope_args
        }

        try:
            results = worker(args)
        except ApiException as err:
            error_body_json = err and hasattr(err, "body") and err.body
            # We can pull the DocuSign error code and message from the response body
            error_body = json.loads(error_body_json)
            error_code = error_body and "errorCode" in error_body and error_body["errorCode"]
            error_message = error_body and "message" in error_body and error_body["message"]
            # In production, you may want to provide customized error messages and
            # remediation advice to the user
            return render_template("error.html",
                                   err=err,
                                   error_code=error_code,
                                   error_message=error_message
                                   )
        if results:
            session["envelope_id"] = results["envelope_id"] # Save for use by other examples
                                                            # that need an envelope ID
            # Redirect the user to the signing ceremony
            # Don't use an iframe!
            # State can be stored/recovered using the framework's session or a
            # query parameter on the return URL (see the makeRecipientViewRequest method)
            return redirect(results["redirect_url"])

    else:
        flash("Sorry, you need to re-authenticate.")
        # We could store the parameters of the requested operation so it could be restarted
        # automatically. But since it should be rare to have a token issue here,
        # we'll make the user re-enter the form data after authentication.
        session["eg"] = url_for(eg)
        return redirect(url_for("ds_must_authenticate"))

# ***DS.snippet.0.start
def worker(args):
    """
    1. Create the envelope request object
    2. Send the envelope
    3. Create the Recipient View request object
    4. Obtain the recipient_view_url for the signing ceremony
    """
    envelope_args = args["envelope_args"]
    # 1. Create the envelope request object
    envelope_definition = make_envelope(envelope_args)

    # 2. call Envelopes::create API method
    # Exceptions will be caught by the calling function
    api_client = ApiClient()
    api_client.host = args["base_path"]
    api_client.set_default_header("Authorization", "Bearer " + args["ds_access_token"])

    envelopes_api = EnvelopesApi(api_client)
    results = envelopes_api.create_envelope(args["account_id"], envelope_definition=envelope_definition)
    
    envelope_id = results.envelope_id
    app.logger.info(f"Envelope was created. EnvelopeId {envelope_id}")

    # 3. Create the RecipientViewRequest object
    recipient_view_request = RecipientViewRequest(
        authentication_method = authentication_method,
        client_user_id = envelope_args["signer_client_id"],
        recipient_id = "1",
        return_url = envelope_args["ds_return_url"],
        user_name = envelope_args["signer_name"], email = envelope_args["signer_email"]
    )
    # 4. Obtain the recipient view URL for the signing ceremony
    # Exceptions will be caught by the calling function
    results = envelopes_api.create_recipient_view(args["account_id"], envelope_id,
        recipient_view_request = recipient_view_request)

    return {"envelope_id": envelope_id, "redirect_url": results.url}

def make_envelope(args):
    """
    Creates envelope
    args -- parameters for the envelope:
    signer_email, signer_name, signer_client_id
    returns an envelope definition
    """

    # Document 1 (PDF) has tag /sn1/
    #
    # The envelope has one recipient:
    # recipient 1 - signer
    with open(path.join(demo_docs_path, ds_config.DS_CONFIG["doc_salary_docx"]), "rb") as file:
        content_bytes = file.read()
    base64_file_content = base64.b64encode(content_bytes).decode("ascii")

    # Create the document model
    document = Document( # Create the DocuSign document object
        document_base64 = base64_file_content,
        name = "Lorem Ipsum", # Can be different from the actual filename
        file_extension = "docx", # Many different document types are accepted
        document_id = 1 # A label used to reference the doc
    )

    # Create the signer recipient model
    signer = Signer( # The signer
        email = args["signer_email"], name = args["signer_name"],
        recipient_id = "1", routing_order = "1",
        # Setting the client_user_id marks the signer as embedded
        client_user_id = args["signer_client_id"]
    )

    # Create a SignHere tab (field on the document)
    sign_here = SignHere( # DocuSign SignHere field/tab
        anchor_string = "/sn1/", anchor_units = "pixels",
        anchor_y_offset = "10", anchor_x_offset = "20"
    )

    text_legal = Text( 
        anchor_string = "/legal/", anchor_units = "pixels", 
        anchor_y_offset = "-9", anchor_x_offset = "5", 
        font = "helvetica", font_size = "size11", 
        bold = "true", value = args["signer_name"], 
        locked = "false", tab_id = "legal_name", 
        tab_label = "Legal name" )

    text_familar = Text( 
        anchor_string = "/familiar/", anchor_units = "pixels", 
        anchor_y_offset = "-9", anchor_x_offset = "5", 
        font = "helvetica", font_size = "size11", 
        bold = "true", value = args["signer_name"], 
        locked = "false", tab_id = "familar_name", 
        tab_label = "Familiar name" )

    salary = 123000

    text_salary = Text(
                anchor_string = "/salary/",
                anchor_units = "pixels",
                anchor_y_offset = "-9",
                anchor_x_offset = "5",
                font = "helvetica",
                font_size = "size11",
                bold = "true",
                value = "${:.2f}".format(salary),
                locked = "true",
                tab_id = "salary",
                tab_label = "Salary")
            
    salary_custom_field = TextCustomField(
        name = "salary",
        required = "false",
        show = "true", # Yes, include in the CoC
        value = str(salary)
    )
    cf = CustomFields( text_custom_fields= [ salary_custom_field ] )
    # Add the tabs model (including the SignHere tab) to the signer
    # The Tabs object wants arrays of the different field/tab types
    signer.tabs = Tabs(sign_here_tabs = [sign_here], text_tabs = [text_legal, text_familar, text_salary])

    # Create the top level envelope definition and populate it
    envelope_definition = EnvelopeDefinition(
        email_subject = "Please sign this document sent from the Python SDK",
        documents = [document],
        # The Recipients object wants arrays for each recipient type
        recipients = Recipients(signers = [signer]),
        custom_fields= cf,
        status = "sent" # Requests that the envelope be created and sent
    )

    return envelope_definition
# ***DS.snippet.0.end

def get_controller():
    """responds with the form for the example"""

    if views.ds_token_ok():
        return render_template("eg016_set_tab_values.html",
                               title="SetTabValues",
                               source_file=path.basename(__file__),
                               source_url=ds_config.DS_CONFIG["github_example_url"] + path.basename(__file__),
                               documentation=ds_config.DS_CONFIG["documentation"] + eg,
                               show_doc=ds_config.DS_CONFIG["documentation"],
                               signer_name=ds_config.DS_CONFIG["signer_name"],
                               signer_email=ds_config.DS_CONFIG["signer_email"]
        )
    else:
        # Save the current operation so it will be resumed after authentication
        session["eg"] = url_for(eg)
        return redirect(url_for("ds_must_authenticate"))